src/main/kotlin/io/libp2p/etc/util/netty/SplitEncoder.kt

```diff
@@ -0,0 +1,20 @@
+package io.libp2p.etc.util.netty
+
+import io.netty.buffer.ByteBuf
+import io.netty.channel.ChannelHandlerContext
+import io.netty.handler.codec.MessageToMessageEncoder
+import kotlin.math.min
+
+/**
+ * Splits large outbound buffers onto smaller chunks of specified max size
+ */
+class SplitEncoder(val maxSize: Int) : MessageToMessageEncoder<ByteBuf>() {
+    override fun encode(ctx: ChannelHandlerContext, msg: ByteBuf, out: MutableList<Any>) {
+        var startIdx = msg.readerIndex()
+        while (startIdx < msg.writerIndex()) {
+            val endIndex = min(startIdx + maxSize, msg.writerIndex())
+            out += msg.retainedSlice(startIdx, endIndex - startIdx)
+            startIdx += maxSize
```

Comment:
Should you not be incrementing `startIdx` here by `endIndex - startIndex` ?

Reply:
`endIndex - startIndex` would differ from `maxSize` on the terminal iteration only when the loop exit